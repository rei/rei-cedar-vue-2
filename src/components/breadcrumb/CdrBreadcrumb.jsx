import modifier from '../../mixins/modifier';
import style from './styles/CdrBreadcrumb.scss';

export default {
  name: 'CdrBreadcrumb',
  mixins: [modifier],
  props: {
    /**
     * Required. List of source breadcrumb property objects
     *
     * {
     *   displayText: Breadcrumb Display Text
     *   url: location breadcrumb should navigate on click
     * }
     */
    items: {
      type: Array,
      default: () => [],
      validator: (value) => {
        if (value.length && value.length > 0) {
          for (let i = 0; i < value.length; i += 1) {
            if (!(typeof value[i].item === 'object')) {
              console.error('Breadcrumb items array missing item key at index ', i); // eslint-disable-line no-console
              return false;
            }
            if (!Object.hasOwnProperty.call(value[i].item, 'name')) {
              console.error('Breadcrumb items array is missing item.name value at index ', i); // eslint-disable-line no-console
              return false;
            }
          }
        }
        return true;
      },
    },
    /**
     * Flag to track container width threshold exceeded
     */
    truncationEnabled: {
      type: Boolean,
      default: true,
    },
  },
  data() {
    return {
      shouldTruncate: this.truncationEnabled && this.items.length > 2,
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-breadcrumb';
    },
    truncate() {
      return this.shouldTruncate;
    },
    ellipsis() {
      return this.truncate ? (<li
        onClick={this.handleEllipsisClick}
        class={this.style['cdr-breadcrumb__item']}
        ref="ellipse"
      >
        <button
          aria-expanded="false"
          class={this.style['cdr-breadcrumb__ellipses']}
          aria-label="ellipsis"
        >
          <svg
            class={this.style['cdr-breadcrumb__ellipses-icon']}
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 24 24"
          >
            <title>
              ellipsis
            </title>
            {/* eslint-disable-next-line max-len */}
            <path d="M17.5 22a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zM12 22a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm-5.5 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z" />
          </svg>
        </button>
        <span
          class={this.style['cdr-breadcrumb__delimiter']}
          aria-hidden="true"
        >
          /
        </span>
      </li>) : '';
    },
    listItems() {
      return this.items.map((breadcrumb, index) => {
        const delimiter = index < this.items.length - 1 ? (<span
          class={this.style['cdr-breadcrumb__delimiter']}
          aria-hidden="true"
        >
          /
        </span>) : '';

        return (<li
          class={this.style['cdr-breadcrumb__item']}
          key={breadcrumb.item.id || breadcrumb.item.name.replace(/ /g, '-').toLowerCase()}
          v-show={!this.truncate || (index >= this.items.length - 2)}
        >
          {this.$scopedSlots.link
            ? this.$scopedSlots.link({
              class: this.style['cdr-breadcrumb__link'],
              href: breadcrumb.item.url,
              content: breadcrumb.item.name,
            })
            : (<a
              class={this.style['cdr-breadcrumb__link']}
              href={breadcrumb.item.url}
            >
              { breadcrumb.item.name }
            </a>)
          }
          {delimiter}
        </li>);
      });
    },
  },
  methods: {
    handleEllipsisClick() {
      this.shouldTruncate = false;
    },
  },
  render() {
    return (<nav
      ref="container"
      class={this.modifierClass}
      aria-label="Breadcrumb"
    >
      <ol
        ref="cdrBreadcrumbList"
        class={this.style['cdr-breadcrumb__list']}
      >
        {this.ellipsis}
        {this.listItems}
      </ol>
    </nav>);
  },
};
