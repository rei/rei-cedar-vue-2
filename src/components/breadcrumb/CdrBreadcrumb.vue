<template>
  <!-- disable lint errors on line length in template -->
  <!-- eslint-disable max-len -->
  <!-- eslint-disable vue/require-v-for-key -->

  <nav
    ref="container"
    :class="modifierClass"
    aria-label="Breadcrumb">
    <ol
      ref="cdrBreadcrumbList"
      :class="$style['cdr-breadcrumb__list']">
      <li
        v-if="truncate"
        @click="shouldTruncate = false"
        :class="$style['cdr-breadcrumb__item']">
        <button
          aria-expanded="false"
          :class="$style['cdr-breadcrumb__ellipses']"
          aria-label="ellipsis">
          <svg
            :class="$style['cdr-breadcrumb__ellipses-icon']"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 24 24">
            <title>
              ellipsis
            </title>
            <path d="M17.5 22a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zM12 22a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm-5.5 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"/>
          </svg>
        </button>
        <span
          :class="$style['cdr-breadcrumb__delimiter']"
          aria-hidden="true">
          /
        </span>
      </li>
      <li
        :class="$style['cdr-breadcrumb__item']"
        v-for="(breadcrumb, index) in items"
        v-show="!truncate || (index >= items.length - 2)"
      >
        <a
          :class="$style['cdr-breadcrumb__link']"
          :href="breadcrumb.item.url">
          {{ breadcrumb.item.name }}
        </a>
        <span
          :class="$style['cdr-breadcrumb__delimiter']"
          aria-hidden="true"
          v-if="index < items.length - 1">
          /
        </span>
      </li>
    </ol>
  </nav>
</template>

<script>
/**
 * Cedar 2 component for breadcrumb
 *
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */

import modifier from 'mixinsdir/modifier';
import breakpoints from 'mixinsdir/breakpoints';
import debounce from 'lodash/debounce';

export default {
  name: 'CdrBreadcrumb',
  mixins: [modifier, breakpoints],
  props: {
    /**
     * Required. List of source breadcrumb property objects
     *
     * {
     *   displayText: Breadcrumb Display Text
     *   url: location breadcrumb should navigate on click
     * }
     */
    items: {
      type: Array,
      default: () => [],
      validator: (value) => {
        if (value.length && value.length > 0) {
          for (let i = 0; i < value.length; i += 1) {
            if (!(typeof value[i].item === 'object')) {
              console.error('Breadcrumb items array missing item key at index ', i); // eslint-disable-line no-console
              return false;
            }
            if (!Object.hasOwnProperty.call(value[i].item, 'name')) {
              console.error('Breadcrumb items array is missing item.name value at index ', i); // eslint-disable-line no-console
              return false;
            }
          }
        }
        return true;
      },
    },
    /**
     * Flag to track container width threshold exceeded
     */
    truncationEnabled: {
      type: Boolean,
      default: true,
    },
    /**
     * Value for max breadcrumb width to container width
     * ratio before truncation is performed
     */
    truncationThreshold: {
      type: Number,
      default: 0.8,
    },
    /**
     * Value for max breadcrumb width to container width
     * ratio before truncation is performed on XS breakpoint
     */
    truncationXSThreshold: {
      type: Number,
      default: 1,
    },
  },
  data() {
    return {
      thresholdExceeded: false,
      shouldTruncate: this.truncationEnabled,
      breadcrumbWidth: 0,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-breadcrumb';
    },
    truncate() {
      return this.shouldTruncate && this.thresholdExceeded;
    },
  },
  mounted() {
    this.breadcrumbWidth = this.getBreadcrumbWidth();
    this.thresholdExceeded = this.calculateTruncation();
    window.addEventListener('resize', debounce(() => {
      this.thresholdExceeded = this.calculateTruncation();
    }, 250));
  },
  methods: {
    getBreadcrumbWidth() {
      const breadcrumbsElements = Array.from(this.$refs.cdrBreadcrumbList.children);
      let totalWidth = 0;
      breadcrumbsElements.forEach((element) => {
        totalWidth += element.offsetWidth || 0;
      });
      return totalWidth;
    },
    calculateTruncation() {
      if (!this.$refs.container) return false;
      const containerWidth = this.$refs.container.offsetWidth || 0;
      const ratio = this.breadcrumbWidth / containerWidth || 0;
      if (this.isXS()) {
        return (ratio > this.truncationXSThreshold);
      }
      return (ratio > this.truncationThreshold);
    },
  },
};
</script>

<style module>
  @import '../../css/settings/_index.pcss';
  @import './styles/CdrBreadcrumb.pcss';
</style>
