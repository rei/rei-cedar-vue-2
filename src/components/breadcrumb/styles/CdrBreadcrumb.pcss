/* ==========================================================================
  # CdrBreadcrumb

  All values should map to variables in
  vars/CdrBreadcrumb.vars.pcss
  in order to allow for theming

  TOC:

    :Base - CdrBreadcrumb
      :Elements
        :container
        :list
        :item
        :link
        :delimiter
        :ellipses
        :ellipses-icon
========================================================================== */

.cdr-breadcrumb {
  line-height: 1;

  /* List
  ========== */
  &__list {
    display: inline-flex;
    flex-wrap: wrap;
    padding: 0;
    margin: 0;
  }

  /* Item
  ========== */
  &__item {
    display: inline-flex;
    flex: none;
    color: $breadcrumb-text-color;
    padding: 0;

    &:last-child {
      color: $breadcrumb-last-child-text-color;
    }

    /* Hover
    ---------- */
    &:hover,
    &:active,
    &:focus {
      color: $breadcrumb-hover-text-color;
    }
  }

  /* Link
  ---------- */
  &__link {
    @include redwood-display-10;

    text-decoration: none;
    color: inherit; /* inherit color from &__item */

    &:hover,
    &:active,
    &:focus {
      color: inherit;
      text-decoration: underline;
    }
  }

  /* Delimiter
  ========== */
  &__delimiter {
    @include redwood-display-10;

    color: $breadcrumb-delimiter-text-color;
    padding-left: $space-quarter-x;
    padding-right: $space-half-x;
  }

  /* Ellipses
    ========== */
  &__ellipses {
    font-family: inherit;
    font-size: inherit;
    line-height: inherit;
    border: none;
    background: none;
    background-color: transparent;
    cursor: pointer;
    display: inline-block;
    padding: 6px 11px; /* extend clickable area */
    margin: -6px -11px; /* preserve layout */

    /* Ellipses Icon
    ========== */
    .cdr-breadcrumb__ellipses-icon {
      border-bottom: 1px solid transparent;
      width: 1.4rem;
      height: 1.4rem;
      fill: $breadcrumb-text-color;
    }

    &:hover,
    &:active,
    &:focus {
      .cdr-breadcrumb__ellipses-icon {
        border-bottom: 1px solid $breadcrumb-hover-text-color;
        fill: $breadcrumb-hover-text-color;
      }
    }
  }

  @media (--md-mq) {
    &__link,
    &__delimiter,
    &__ellipses {
      @include redwood-display-20;
    }

    &__ellipses .cdr-breadcrumb__ellipses-icon {
      width: 1.6rem;
      height: 1.6rem;
    }
  }
}
