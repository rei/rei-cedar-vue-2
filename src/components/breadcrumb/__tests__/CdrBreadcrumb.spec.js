import { shallowMount, mount } from '@vue/test-utils';
import CdrBreadcrumb from 'componentdir/breadcrumb/CdrBreadcrumb';

describe('CdrBreadcrumb', () => {
  const BreadcrumbItems = [
    {
      item: {
        url: 'http://google.com',
        name: 'Breadcrumb Step 1',
      },
    },
    {
      item:{
        url: 'http://rei.com',
        name: 'Long Breadcrumb Step 2',
      },
    },
    {
      item:{
        url: 'http://yahoo.com',
        name: 'Breadcrumb Step 3',
      },
    },
    {
      item:{
        url: 'http://bing.com',
        name: 'Really Really Long Breadcrumb Step 4',
      },
    },
    {
      item:{
        url: 'http://amazon.com',
        name: 'Last Step',
      },
    },
  ];

  test('renders correctly', () => {
    const wrapper = mount(CdrBreadcrumb);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('breadcrumb should not truncate with fewer than 3 items', () => {
    const items = [
      {
        item: {
          url: 'http://google.com',
          name: 'Breadcrumb Step 1',
        },
      },
      {
        item:{
          url: 'http://rei.com',
          name: 'Long Breadcrumb Step 2',
        },
      },
    ];
    const wrapper = shallowMount(CdrBreadcrumb, {
      propsData: {
        items: items,
      }
    });
    expect(wrapper.vm.truncate).toBe(false);
  });

  it('breadcrumb should truncate with 3 or more items', () => {
    const items = [
      {
        item: {
          url: 'http://google.com',
          name: 'Breadcrumb Step 1',
        },
      },
      {
        item:{
          url: 'http://rei.com',
          name: 'Long Breadcrumb Step 2',
        },
      },
      {
        item:{
          url: 'http://yahoo.com',
          name: 'Breadcrumb Step 3',
        },
      },
    ];
    const wrapper = shallowMount(CdrBreadcrumb, {
      propsData: {
        items: items,
      }
    });
    expect(wrapper.vm.truncate).toBe(true);
    wrapper.find({ref: 'ellipse'}).trigger('click');
    expect(wrapper.vm.truncate).toBeFalsy();

  });

  it('breadcrumb link can be overridden with link scopedSlot', () => {
    const items = [
      {
        item: {
          url: 'http://rei.com',
          name: 'Scoped',
        },
      },
    ];
    const wrapper = shallowMount(CdrBreadcrumb, {
      propsData: {
        items: items,
      },
      scopedSlots: {
        link: '<p slot-scope="link">{{link.href}} TEST {{link.content}} {{link.class}}</p>'
      }
    });
    expect(wrapper.text()).toBe('http://rei.com TEST Scoped cdr-breadcrumb__link');
  });

});
