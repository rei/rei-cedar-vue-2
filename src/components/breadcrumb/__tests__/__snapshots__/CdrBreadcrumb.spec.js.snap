// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrBreadcrumb renders correctly 1`] = `
<nav
  aria-label="Breadcrumb"
  class="cdr-breadcrumb"
>
  <ol
    class="cdr-breadcrumb__list"
  />
</nav>
`;
